### PR TITLE
feat(desktop): consolidate AI Clone

### PR DESCRIPTION
**Pull request by undivisible**
## Summary

- Consolidates the AI Clone work into one cross-platform implementation for macOS and Windows.
- Keeps Beeper access local and requires approval for every reply; legacy auto configurations downgrade to drafts.
- Reuses the current desktop shells while adding the bridge and test coverage needed to exercise the feature.

## Product invariants affected

- INV-AUTH-1
- INV-CHAT-1

## Validation

- `xcrun swift build -c debug --package-path Desktop`
- `bun node_modules/vitest/vitest.mjs run src/main/aiClone`
- `bun node_modules/typescript/bin/tsc --noEmit -p tsconfig.node.json --composite false`
- `bun node_modules/typescript/bin/tsc --noEmit -p tsconfig.web.json --composite false`
- `bash scripts/dev-harness/run-python.sh .github/scripts/pr_preflight.py --lane local --base upstream/main --pr-body-file /tmp/omi-ai-clone-pr.md`

## Known local test blocker

`xcrun swift test --package-path Desktop --filter AIClone` reaches the unrelated test target and is blocked by existing Xcode Beta errors for three `URLProtocol` test stubs inheriting an unavailable `Sendable` conformance.

<a href="https://cubic.dev/pr/BasedHardware/omi/pull/10933?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

Failure-Class: none

Line-Count-Exception: desktop/macos/Desktop/Sources/OmiApp.swift | 1590 -> 1594 | AI Clone auto-resume hook adds four lines on an already-oversized app entry file

---

> [!NOTE]
> **High Risk**
> The feature can send or draft real messages on the user's behalf, stores third-party API tokens, and runs LLM policy on untrusted inbound chat—macOS Auto mode adds automatic sends despite benchmarks and rate limits.
> 
> **Overview**
> Introduces **AI Clone** on macOS and Windows so Omi can draft or send chat replies through **Beeper Desktop** (WhatsApp, Telegram, iMessage, etc.) using persona and memories, with Beeper tokens kept local.
> 
> On **macOS**, new `AIClone` modules add a `BeeperDesktopClient`, a WebSocket listen loop, an LLM reply engine with injection/sensitive-content policy, per-chat **Off / Draft / Ask / Auto** (Auto requires a history-replay **benchmark** and confidence caps), approvals UI in Settings, launch **resume** when enabled, and automation actions (`ai_clone_*`). The automation bridge registers these via `registerAICloneAndRewindActions()`.
> 
> On **Windows**, a main-process `AiCloneService` connects over IPC, generates replies via **`/v2/messages`** (with desktop completions fallback), and is **draft/approval-only**—legacy `auto` modes are stored/served as **draft**. Beeper tokens use encrypted persistence; a per-chat queue avoids dropping rapid messages.
> 
> CI adds an **ai-clone** e2e flow, extends desktop contract path triggers, and documents the feature in the Windows README.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit cabc566e205329aaf78d0775e34bc58be2f5a23c. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by Git-on-my-level** [comment]
Thanks for consolidating this — the cross-platform direction and the fail-closed defaults are promising, especially keeping Beeper local, defaulting chats to off/draft, and adding Windows unit coverage for the reply queue.

I’m going to keep this as human-review/changes-needed for now because there are a couple of blockers before maintainers can evaluate the product side safely:

- The macOS Swift CI is currently red. The failure is in the new AI Clone Swift path: `BeeperCursorPage<BeeperChat>` / `BeeperCursorPage<BeeperMessage>` are crossing actor isolation without being `Sendable` in `AICloneService.swift`, so the desktop Swift test/build lane cannot pass yet.
- The macOS live-event loop can still silently drop messages for an enabled chat while a previous reply is in flight. In `AICloneService.handleLiveEvent`, the code claims `inbound.id` as processed before checking `inFlightChatIDs`; if another message arrives during generation, it returns at the in-flight guard with that message already marked handled. The Windows implementation explicitly queues/parks this case, so macOS should get the same “newest pending message runs after the current one” behavior (and a regression test) before this can be considered safe.
- This feature can draft or auto-send personal chat replies through Beeper, stores a Beeper access token, and adds automation/workflow coverage, so it needs human maintainer/product/security review after the build and event-loop blocker are fixed. The current scope is cohesive around one desktop feature; I’m not asking for a split, just a clean CI run plus focused validation of the reply/send safety model.

No formal approval from automation here because the PR is draft, CI is failing, and this is a sensitive desktop messaging/autonomy feature.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for the quick update. The macOS in-flight message-drop issue I called out earlier looks addressed now: `AICloneInboundQueue` parks the newest pending inbound message and the new regression coverage exercises that path.

I’m still going to keep this changes-needed / human-review-required for a few reasons before maintainers evaluate the product side:

- CI is still not green. The current `Hygiene` job is failing in the release-process guard, and the desktop Swift static/test contract is still in progress on this head. Please get the required lanes green before this is considered merge-ready.
- Windows can auto-send for any per-chat `auto` setting after only a confirmation dialog; unlike the macOS path, there is no replay benchmark/confidence gate before enabling or sending in Auto. For a feature that can speak in the user’s personal chats, Windows should either implement the same safety gate or clearly downgrade to reviewable drafts until equivalent evidence exists.
- On Windows, `handleIncoming` marks a message processed before `respond()` has actually produced/saved/sent anything. If the app resumed with AI Clone enabled but main has no Firebase token yet (or reply generation fails before a draft is saved), that inbound message is recorded as handled and duplicate/upsert events will not retry it. Please mark processed only after a draft/send is successfully recorded, or add a retry/parked failure state with a regression test.

The scope is still cohesive around one desktop AI Clone feature, so I’m not asking for a split. Because this touches personal messaging, Beeper token handling, and auto-send behavior, it should remain under human product/security review after the safety and CI issues are fixed.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by Git-on-my-level** [comment]
Thanks for the follow-up fixes. I re-reviewed the current head and the previous automation blockers look resolved:

- Windows no longer exposes Auto mode in the AI Clone UI; the Windows responder now always creates reviewable drafts, so every send requires user approval.
- Windows now marks an inbound message processed only after a draft is actually saved, with regression coverage for the no-token/failure retry path.
- GitHub checks are currently green, and I removed the stale `needs-tests` label.

I also dismissed the stale automation change-request review tied to the prior head. This still should not be treated as automation approval: the PR is draft and this is a sensitive desktop messaging feature touching Beeper token handling, personal chat content, and user-impersonating reply generation, so please keep `security-review` / `needs-maintainer-review` until a human maintainer signs off on the product and safety model.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for keeping this cohesive and for getting the focused AI Clone checks green. I re-reviewed the current head and found one remaining docs/product-safety mismatch that should be fixed before maintainer sign-off:

- `desktop/windows/README.md` still documents Windows AI Clone as offering **Auto** mode that "sends immediately" per chat. The current Windows UI only exposes `Off` and `Draft`, and the Windows responder always returns a reviewable draft even for legacy `auto` settings. That implementation direction is safer and matches the prior blocker resolution, but the committed README now over-promises an auto-send behavior that Windows intentionally does not expose. Please update the Windows docs to say this Windows version is draft/approval-only for now, or implement and validate an equivalent safety gate before documenting Auto.

I verified the focused Windows AI Clone tests locally (`npx vitest run src/main/aiClone src/shared/omiSse.test.ts`: 7 files / 40 tests passed) and GitHub checks are currently green. I am not asking for a split; this is a cohesive desktop AI Clone PR. Because it still touches Beeper token handling, personal chat content, and user-impersonating replies, please keep the human product/security review labels until a maintainer signs off.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by Git-on-my-level** [comment]
Thanks for the follow-up — I rechecked the current head and the previous Windows docs blocker is resolved. The README now accurately says the Windows AI Clone path is draft/approval-only, and legacy `Auto` settings still create a reviewable draft rather than sending automatically.

I also checked the Windows README section that mentions coding-agent setup. This change does not add or change repo agent instructions or review policy; it preserves the existing user-facing ACP setup guidance, so I don’t see a new agent-behavior risk from that documentation.

I’m removing the stale `docs-accuracy` label and dismissing the stale automation change request tied to the older README mismatch. I’m still leaving this for human maintainer/product/security review before merge because it is a draft PR and the feature touches Beeper access-token handling, personal chat content, and user-impersonating reply generation. No split requested; this remains a cohesive desktop AI Clone change.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by undivisible** [comment]
Rebased onto current `main` — the PR is mergeable again.

Two things changed on this head:
- Merge conflicts resolved in `SettingsSidebar.swift` (kept main's extracted `SettingsSidebarRoutes.visibleSections`, re-added `.aiClone`), `preload/index.ts`, and `routes/manifest.ts`.
- `AIClonePage.swift` ported off the `OmiColors` palette that main deleted, onto `Ink` / `PageGlass`. Without this the merged tree did not compile.

The Windows README already reads draft/approval-only, so no change was needed there.

Verified locally: `make preflight` — 24 checks passed; `npm run typecheck` (node + web) clean; `npx vitest run src/main/aiClone src/shared/omiSse.test.ts src/renderer/src/routes/manifest.test.ts` — 8 files / 72 tests passed.

One thing to know: `xcrun swift build -c debug --package-path Desktop` fails locally on `DesktopHomeView.swift:454` ("unable to type-check this expression in reasonable time"). I reproduced that on a clean `origin/main` worktree, so it is my local Xcode Beta toolchain, not this branch. CI's pinned Xcode is the authority — leaving this in draft until the Swift lane reports green.

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for continuing to consolidate this. The Windows side looks much safer on this head: `desktop/windows/src/main/aiClone/responder.ts` and `desktop/windows/src/main/aiClone/store.ts` both downgrade legacy `auto` settings to draft-only, `desktop/windows/src/main/aiClone/service.ts` only marks a message processed after a draft is successfully created, and the new tests cover the Beeper client, queueing, responder rules, store persistence, SSE parsing, and route manifest. The workflow path update in `.github/workflows/desktop-backend-contracts.yml` also correctly brings the macOS AI Clone automation action source into the desktop contract checks.

I do need one macOS lifecycle fix before this can merge:

- `desktop/macos/Desktop/Sources/AIClone/AICloneService.swift`: `handleLiveEvent` calls `claimInbound(inbound.id)` before queueing/processing the message (lines 293-299). If `processInbound` then fails for a transient reason — Beeper list/send error, backend/model failure, missing owner identity, parsing failure, etc. — the catch block only records a failed activity entry, but the message id remains in `processedInboundMessageIDs`. A duplicate `message.upserted` for that same message (or a reconnect replay) will be suppressed forever, so the clone can silently drop an otherwise actionable inbound message after one temporary failure. This regresses the reliability guarantee that the Windows path now has in `desktop/windows/src/main/aiClone/service.ts`, where accepted messages are marked processed only after `respond()` succeeds. Please move the macOS claim/mark-processed step to after a successful terminal action, or unclaim/retry-safe-fail when `processInbound` throws, and add a Swift regression test for “first reply attempt fails, duplicate event retries instead of being ignored.”

Other file-specific observations from this pass:

- `desktop/macos/Desktop/Sources/AIClone/AICloneModels.swift` has the right high-level safety shape for Auto mode: benchmark gating, confidence downgrade to draft, sensitive-content downgrade, and injection decline. That makes the retry bug above more important, because it is in the delivery lifecycle rather than the policy ladder.
- `desktop/macos/Desktop/Sources/AIClone/AICloneReplyEngine.swift` correctly treats incoming chat as untrusted data and returns structured verdicts, but a parse/transport failure currently feeds into the same permanent-claim path in `AICloneService.swift`.
- `desktop/windows/src/main/aiClone/replyEngine.ts` reuses the shared SSE parser and avoids sending plan-limit notices as chat replies; the `/v2/messages` fallback is still a product/security-sensitive surface, so maintainer sign-off is appropriate even after the macOS bug is fixed.
- `desktop/windows/README.md` now describes the Windows AI Clone behavior as draft/approval-only and updates the coding-agents setup section. That coding-agent documentation can influence how users wire Claude/OpenClaw/Hermes/Codex, but I don’t see unsafe repo-agent instructions in it; it is still worth maintainer review because it changes public guidance for agent setup.

I ran the focused Windows validation in an isolated worktree with no secrets: `pnpm run typecheck` and the focused Vitest set for `src/main/aiClone/*.test.ts`, `src/shared/omiSse.test.ts`, and `src/renderer/src/routes/manifest.test.ts`; all passed (72 tests). I did not run the macOS Swift suite locally in this Linux environment, so the macOS retry concern is from static review of the changed Swift code path.

Leaving this for product/security-sensitive desktop maintainer review after the macOS retry/drop fix.

— Automated maintainer review by Omi PR monitor (`glm-5.2`)

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by undivisible** [comment]
Worked through every open Codex and cubic inline finding on this head. Disposition below; fixes are in the commits pushed on top of `498b7853`.

**Fixed**

| Finding | Change |
|---|---|
| Inbound claim never released on failure (P1, also @Git-on-my-level's blocker) | `releaseInbound` on any non-terminal outcome, so redelivery retries |
| Verdict acted on under stale policy after the awaits | Policy re-read before `perform`; stale verdict discarded and claim released |
| Approval lost on blank edit / failed send; Clear discarded approvals | Blank check first, failed send re-queues the edited text, Clear touches the log only |
| Beeper page order assumed | Explicit oldest-first sort before `threadLines`, the benchmark, and the Windows transcript |
| `server.base_url` from discovery trusted | Adopted only when scheme is http(s) and host is loopback |
| Out-of-range `confidence` satisfied the auto-send gate | Value and threshold must both be finite and in 0…1 |
| Declined sensitive message logged as Drafted/Asked | Silence guard now runs before safety mapping |
| Injection-flagged benchmark samples judged as matches | Counted as a miss |
| Listener not resumed after restart | `resumeIfEnabled()` from `applicationDidFinishLaunching` |
| Backend `error:` frames staged as reply content | Dropped in the shared SSE parser |
| Windows: single hard-coded Beeper port, no request deadline | `/v1/info` probe across both ports, bounded `AbortSignal` |
| Windows: valid-JSON-but-malformed store crashed startup | Root and collections validated at the read boundary |
| Windows: undated messages bypassed the session cutoff | Ignored, matching macOS |
| Windows: work parked/in-flight when the clone was turned off | Generation stamp retires it before it can persist a draft |
| Windows: auth not rebound on sign-out or account switch | Credential cleared on sign-out, retired and rebound on every account change |
| Ungrounded completions fallback emitted no telemetry | Routed through the shared `recordFallback` |
| Mid-stream failure rejected instead of returning a result | Returns `network` |
| Plus: failed `setChatMode` handling, status copy reflecting `enabled`, dead `getChatModes()`, `result.detail.enabled` in the e2e flow, `AiClone` mock in the manifest test, `Authorization: Bearer` asserted in the Beeper transport tests |

**Not changed, with reasons**

- **`GET /v1/chats` is not a 404.** The Beeper reference documents both *List chats* (`GET /v1/chats`) and *Search chats* (`GET /v1/chats/search`); the Windows probe uses the former deliberately. Same source documents port 23373 as current — the Windows client now probes 23374 as well for robustness, not because 23373 is retired.
- **Windows clone generation still posts to `/v2/messages`.** Real concern (persists a chat row, counts against quota, runs goal extraction), but the fix is a stateless owner-authenticated completion surface, not a patch here. Needs a maintainer decision — flagged, not silently deferred.
- **Auto-send not removed.** That is a product decision, not a defect; this PR already carries `security-review` / `needs-maintainer-review` for exactly that surface.
- **Startup token wait.** A message received before the renderer supplies a token is *not* marked processed, so redelivery retries it. The existing regression test pins that contract; adding a blocking wait would change it without evidence it is needed.
- **Preferring the `done:` payload over streamed text.** No evidence in the backend that the terminal frame is citation-stripped relative to the stream; left alone rather than guessing.

Verification: `npx vitest run src/main/aiClone src/shared/omiSse.test.ts src/renderer/src/routes/manifest.test.ts` — 81 passed. Each behavior fix was mutation-checked: reverting it fails its own new test. `npm run typecheck` clean. `make preflight` passes. macOS Swift build and tests are on CI (local Xcode 27 hits a pre-existing type-check timeout in `DesktopHomeView.swift`, which this branch does not touch).

File: desktop/windows/src/main/aiClone/beeperClient.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+      return { ok: true, value: (await res.json()) as T }
+    } catch {
+      return { ok: true, value: undefined as T }
+    }
+  }
+
+  /** Cheap probe used by Connect and by reachability checks. */
+  async validateToken(): Promise<BeeperResult<void>> {
+    const r = await this.request<unknown>('GET', '/v1/chats')
+    return r.ok ? { ok: true, value: undefined } : r
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/windows/src/main/aiClone/beeperClient.ts:144]: Not a defect: the Beeper reference documents both *List chats* (`GET /v1/chats`) and *Search chats* (`GET /v1/chats/search`). The probe uses List deliberately; no change.

File: desktop/windows/src/main/aiClone/beeperClient.ts [hunk 2]
```diff
@@ -0,0 +1,187 @@
+// Thin client for the Beeper Desktop API — a localhost REST + WebSocket server
+// that Beeper Desktop runs on port 23373. The user creates an access token in
+// Beeper → Settings → Integrations ("Approved connections"); everything stays
+// on-device. Docs: https://developers.beeper.com/desktop-api
+import WebSocket from 'ws'
+
+export const BEEPER_BASE_URL = 'http://localhost:23373'
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/windows/src/main/aiClone/beeperClient.ts:9]: Windows now probes `/v1/info` on 23373 and 23374 and adopts whichever answers. Note the published reference still shows 23373 as current, so this is robustness rather than a retired port.

File: desktop/windows/src/main/aiClone/replyEngine.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  let res: Response
+  try {
+    res = await doFetch(`${args.apiBase}/v2/messages`, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+        Authorization: `Bearer ${args.firebaseToken}`
+      },
+      body: JSON.stringify({ text: prompt })
+    })
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/windows/src/main/aiClone/replyEngine.ts:97]: Real, not fixed here: routing clone generation off the persisted default-chat lane needs a stateless owner-authenticated completion surface. Flagged for a maintainer decision rather than patched in this PR.

File: desktop/macos/Desktop/Sources/AIClone/AICloneService.swift [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+            chatTitle: chatTitle,
+            network: network,
+            inboundPreview: Self.preview(inbound.text),
+            replyText: reply,
+            confidence: decision.confidence,
+            receivedAt: Date()))
+      }
+    case .sentAutomatically:
+      if let reply = decision.reply {
+        _ = try await client.sendMessage(chatID: chatID, text: reply, replyToMessageID: inbound.id)
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/macos/Desktop/Sources/AIClone/AICloneService.swift:492]: Auto-send is a product decision, not a defect. This PR already carries `security-review` / `needs-maintainer-review` for that surface; leaving it to the maintainer.

File: desktop/windows/src/main/aiClone/service.ts [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+    if (this.processed.size > 2_000) {
+      // Drop the oldest half (Set iterates in insertion order).
+      const keep = [...this.processed].slice(-1_000)
+      this.processed = new Set(keep)
+    }
+  }
+
+  private async respond(chat: BeeperChat, message: BeeperMessage): Promise<boolean> {
+    const chatTitle = chat.title ?? chat.id
+    if (!this.firebaseToken) {
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/windows/src/main/aiClone/service.ts:283]: Not changed: a message received before the token arrives is not marked processed, so redelivery retries it. The existing regression test pins that contract; a blocking wait would change it without evidence it is needed.

File: desktop/windows/src/main/aiClone/replyEngine.ts [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    const { done, value } = await reader.read()
+    if (done) break
+    const chunk = decoder.decode(value, { stream: true })
+    raw += chunk
+    acc.feed(chunk)
+  }
+  acc.end()
+
+  const text = cleanReply(acc.text)
+  if (text) return { ok: true, text }
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/windows/src/main/aiClone/replyEngine.ts:125]: Not changed: no evidence in the backend that the terminal `done:` text is citation-stripped relative to the stream. Left alone rather than guessing.

**Review by Git-on-my-level — COMMENTED** [review]
Thanks for the update — the blocker from my previous pass looks resolved on this head. In `desktop/macos/Desktop/Sources/AIClone/AICloneService.swift`, the transient-failure path now calls `releaseInbound(inbound.id)` when `processInbound` throws, and `desktop/macos/Desktop/Tests/BeeperDesktopClientTests.swift` adds the regression coverage proving a released inbound message can be claimed again instead of being dropped forever.

A few file-specific notes from this pass:

- `desktop/macos/Desktop/Sources/AIClone/AICloneModels.swift` keeps the macOS trust ladder fail-closed: malformed confidence downgrades to draft, sensitive/injection-like content does not auto-send, and Auto remains benchmark-gated.
- `desktop/macos/Desktop/Sources/AIClone/AICloneReplyEngine.swift` explicitly frames Beeper chat content as untrusted data and requires a structured JSON verdict; that is the right shape for this feature’s prompt-injection surface.
- `desktop/macos/Desktop/Sources/AIClone/BeeperDesktopClient.swift` only adopts Beeper’s advertised API base URL when it is loopback, so the Beeper bearer token is not redirected off-box by a malicious or squatting local service response.
- `.github/workflows/desktop-backend-contracts.yml` now includes `desktop/macos/Desktop/Sources/AIClone/AICloneAutomationActions.swift` in the desktop backend contract trigger set, which is appropriate for the new automation bridge actions.
- `desktop/windows/src/main/aiClone/responder.ts`, `desktop/windows/src/main/aiClone/store.ts`, and `desktop/windows/src/main/aiClone/service.ts` keep the Windows implementation draft/approval-only: legacy `auto` modes are downgraded, accepted messages are marked processed only after a draft is created, and in-flight work is retired when listening is disabled.
- `desktop/windows/src/main/aiClone/replyEngine.ts` and `desktop/windows/src/shared/omiSse.ts` avoid turning SSE `error:`/`done:` service frames into contact-visible reply text, and they keep fallback to the desktop completions lane explicit.
- `desktop/windows/src/renderer/src/pages/AiClone.tsx` clears and rebinds the Firebase token on auth changes, which avoids continuing to draft from a previous signed-in user’s session.
- `desktop/windows/README.md` changes public AI Clone and coding-agent setup guidance. That can influence how users configure external AI coding agents in this repo, but I don’t see unsafe repo-agent instructions in the added text; it accurately says Windows AI Clone is draft/approval-only and that external agents use an allowlisted environment.

Validation I could run here:

- `./node_modules/.bin/tsc --noEmit -p tsconfig.node.json --composite false`
- `./node_modules/.bin/tsc --noEmit -p tsconfig.web.json --composite false`
- `./node_modules/.bin/vitest run src/main/aiClone/*.test.ts src/shared/omiSse.test.ts src/renderer/src/routes/manifest.test.ts` — 8 files / 81 tests passed.

I could not run the macOS Swift suite from this Linux review environment. Also note that `pnpm run typecheck`/`pnpm exec ...` tried to invoke `pnpm install`, and the existing `verify-pimono-unpack` postinstall check failed on unresolved packaged-agent closure packages before the direct `node_modules/.bin/*` commands succeeded; that looks like an environment/package-install issue rather than a regression in the changed AI Clone files.

Given the feature can draft/send messages through Beeper, stores a local Beeper token, touches workflow coverage, and needs a rebase/conflict resolution, this still needs human desktop/security maintainer sign-off before merge. But the specific stale macOS retry/drop change request is resolved on this head.

— Automated maintainer review by Omi PR monitor (`glm-5.2`)

---
*by AI on behalf of David — product/security-sensitive desktop messaging automation still needs human maintainer sign-off before merge.*

**Comment by Git-on-my-level** [comment]
Thanks for the latest head. I rechecked the current diff and I’m keeping this non-blocking / human-review-required rather than approving, mainly because this is product/security-sensitive desktop messaging automation and the current required desktop Swift gate is still not green (`Desktop Swift Build & Tests` is failing because the static/test contract result is `cancelled`).

File-specific notes from this pass:

- `desktop/macos/Desktop/Sources/AIClone/AICloneService.swift` now resumes an enabled clone from app launch via `resumeIfEnabled()`, releases claimed inbound IDs on transient reply-pipeline failure with `releaseInbound(inbound.id)`, and re-reads enabled/chat-mode state before acting on a delayed verdict. Those are the right lifecycle guards for a feature that can speak through Beeper.
- `desktop/macos/Desktop/Sources/AIClone/AICloneModels.swift` keeps Auto gated by a benchmark result and finite confidence; sensitive/injection-like content downgrades to draft/approval instead of auto-send.
- `desktop/macos/Desktop/Sources/AIClone/AICloneReplyEngine.swift` treats incoming Beeper message text as untrusted data and requires a structured JSON verdict with `suspected_injection`, which is important for the prompt-injection surface.
- `desktop/macos/Desktop/Sources/AIClone/BeeperDesktopClient.swift` only adopts Beeper’s advertised `/v1/info` base URL when it is loopback, avoiding bearer-token redirect off-box.
- `desktop/macos/Desktop/Sources/AIClone/AICloneAutomationActions.swift`, `DesktopAutomationBridge.swift`, and `.github/workflows/desktop-backend-contracts.yml` add AI Clone bridge actions and include the new action source in the desktop backend contract trigger set; that workflow/automation surface is appropriate to keep under human sign-off.
- `desktop/windows/src/main/aiClone/responder.ts`, `chatTaskQueue.ts`, and `service.ts` keep Windows draft-only: legacy `auto` settings still produce reviewable drafts, per-chat work is serialized/coalesced, and messages are marked processed only after a draft is saved.
- `desktop/windows/src/main/aiClone/store.ts` uses the injected Electron safeStorage/DPAPI path for the Beeper token and sanitizes the JSON store on load, so token/state persistence is scoped and testable.
- `desktop/windows/src/main/aiClone/replyEngine.ts` and `desktop/windows/src/shared/omiSse.ts` avoid treating `error:`, `done:`, `message:`, and `think:` frames as contact-visible reply text, which is the right fail-closed behavior for generated replies.
- `desktop/windows/src/renderer/src/pages/AiClone.tsx`, `preload/index.ts`, `main/ipc/aiClone.ts`, `routes/manifest.ts`, and `shared/types.ts` wire the new page/IPC surface cleanly through typed bridge methods, with route tests covering the nav addition.
- `desktop/windows/README.md` now documents Windows AI Clone as draft/approval-only and includes setup guidance for Beeper plus the existing coding-agent ACP section. This documentation can affect how users configure external coding agents around this repo, but I don’t see new unsafe agent instructions; the maintainer consequence is mainly that the product/docs claim about replying through Beeper should be reviewed for accuracy before release.

I did not find a new concrete code blocker beyond the current CI state, and I’m not asking for a split because the PR is cohesive around one desktop AI Clone feature. This should stay with human desktop/security maintainer review before merge because it can draft/send chat replies through Beeper, stores a local messaging token, uses memory/persona grounding, and touches automation/workflow coverage.

— Automated maintainer review by Omi PR monitor (`glm-5.2`)

---
*by AI on behalf of David — product/security-sensitive desktop messaging automation needs human maintainer sign-off before merge.*
